### ISSUE TITLE
Issue with: react/require-default-props

### ISSUE THREAD

**Issue by jannikkeye** [peerigon/eslint-config-peerigon#68]
The rule leads to an error when optional props do not specify default props. Say I have a component that takes an optional prop A and if A is set I render X if not I render Y. The fact that I am not using `.isRequired` indicates that the prop can be undefined. In those cases I do not need defaultProps. Of course I could set defaultProps to undefined or null but why?

For all the cases where we really want defaultProps the rule is great. I have about 250 errors for this rule alone in the Orderfox project. While there maybe some "real" errors there, most are not.

It doesn't feel right to disable the rule for all those occurrences not writing useless defaultProps definitions. Am I missing something obvious here for those cases I mention?

Can this rule be a warning instead?

**Comment by jhnns** [comment]
I would not use a warning because warnings are also annoying. I used warnings in situations where there might be valid use-cases but 90% of all occurrences its an error. In the valid situations you can disable the warning.

I agree that it would be annoying to set these props as `undefined`.

I need to look into your example. But if there is not a better solution, I would just disable that rule.

**Comment by jannikkeye** [comment]
Take for example this:

```
function LanguageContainer({messages, children, language}) {
    const us = en;
    const mx = es;

    moment.locale(language);

    if (!messages) {
        messages = enLocale;
    }

    const flattenedMessages = flatten(messages);

    Object.keys(flattenedMessages).forEach(key => {
        flattenedMessages[key] = flattenedMessages[key].replace(/{{{/g, "{").replace(/}}}/g, "}");
    });

    return (
        <IntlProvider
            locale={transformLocaleKey(language)}
            messages={flattenedMessages}
        >
            <IntlGlobalProvider>
                { children }
            </IntlGlobalProvider>
        </IntlProvider>
    );
}

LanguageContainer.propTypes = {
    children: PropTypes.any.isRequired,
    language: PropTypes.string.isRequired,
    messages: PropTypes.object,
};
```

Ignore most of it except the check for messages and the prototypes declaration. Messages is not required. I usually do not set default props for cases like this, because I'd have to set either undefined or null, but I don't really have to.

Am I missing a best practice here? Cases like these occur quite often in our components.

**Comment by jhnns** [comment]
I agree that adding `undefined` to `defaultProps` is tedious if you haven't done it yet and you've got a lot of errors now. It makes sense to disable that rule for your project.

Do you think it's also tedious in new React projects (without TypeScript)? If yes, I'm also open for disabling it in our config although I like the explicitness.

Related discussions:
- https://github.com/yannickcr/eslint-plugin-react/issues/1153
- https://github.com/yannickcr/eslint-plugin-react/issues/1987

What's interesting: `defaultProps` is likely to be deprecated for functional components in the future. There's an RFC: https://github.com/reactjs/rfcs/pull/107

**Comment by jannikkeye** [comment]
Well, I cannot share the fairly strict view of the person in those discussions. For me, when in propTypes I omit `.isRequired` its clear enough. I understand that explicitness has its value but at the moment I don't have enough hands on experience of writing explicit default props to form an opinion about those. I can only say from my experience, reading propTypes tells me what can be undefined or have a default value and reading default props tells me which of those previously seen props have a default value. To be honest, I never gave this much thought.

All in all I'd prefer a solution like default parameters for props in the function params or a solution that merges propTypes and default props. A solution where both are immediately obvious.

So in conclusion: I'm gonna disable this rule for now in orderfox and I'll have to form a stronger opinion on either way in another project I guess.

**Comment by jhnns** [comment]
Here's also a good blog post about the difference of `defaultProps` and function default props: https://medium.com/@matanbobi/react-defaultprops-is-dying-whos-the-contender-443c19d9e7f1 (not directly related to this issue, just a good comparison)
